### PR TITLE
Add HTML parsing features

### LINKED ISSUE #2: Add HTML parser/replacer
The original JavaScript Twemoji client allows replacing all emojis in a given text with the corresponding Twemoji image tag in one go.
We should provide a similar method like `Twemoji::parse($html)` which will search for emojis and replace them.

As this requires a DOM/HTML parser to don't replace emojis in image alternate attributes for example this should be an opt-in feature. SO the DOM library shouldn't be part of the default dependencies but the `suggestions`.

The method should also be flagged as experimental so everyone knows that it's possible that this method makes trouble.

To test this we should use snapshot testing on some blog posts for example. (could be faked)
They should cover emojis in plain text `.txt`, in the displayed content of HTML (outside of tags, in tag(s), with body or without) and as part of HTML attributes.

content parser and emoji replacer for:
* [x] Text `.txt` https://github.com/Astrotomic/php-twemoji/issues/3
* [ ] HTML `.html`
* [ ] Markdown `.md`

### PR DESCRIPTION
**Pull request by mallardduck**
This is a WIP PR to address #2 - however I will resubmit/force-push this branch once we have #10 merged in to fix PHP 8.1 support fully.

As it stands now, here are the remaining concerns to address:
1. ~~How to flag as experimental? Is there an PHPDoc annotation I can apply to the class?~~ Might be addressed, waiting for confirmation.
2. Based on the initial issue it sounded like a new method on `Twemoji` but I opted for a new Replacer class. This way the new functionality doesn't directly modify the base Twemoji behaviour. (If this works well for HTML I can work on a follow up PR to address adding a Markdown Replacer class)
3. ~~How should the HTML replacer filter nodes for emoji replacement? I need to do more testing around this and add more test data to the pest datasets. This will help us be more informed on how PHP will filter things out best.~~ Solved: Now using XPath to filter only the `Text` type nodes from the document. In the DOM the [Text](https://developer.mozilla.org/en-US/docs/Web/API/Text) [node type](https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType) represents the inner Text of an element. So by filtering the page first to the body, then selecting all text elements we can safely catch all emoji.

### CODE REVIEW

File: src/HtmlReplacer.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        // Parse the html
+        $parsedHtml = new HtmlPage($html);
+        $body = $parsedHtml->getBody();
+
+        if ($body->children()->count() === 0) {
+            return $html;
+        }
+
+        // Use xpath to filter only the "TextNodes" within each "Element"
+        $textNodes = $body->filterXPath('.//*[normalize-space(text())]');
```
Comment (mallardduck) [inline src/HtmlReplacer.php:36]: This part selects the actual `Text` type nodes within the `body` of the parsed doc.

File: src/HtmlReplacer.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    public function __construct()
+    {
+        if (! class_exists(HtmlPageCrawler::class)) {
+            throw new RuntimeException(
+                sprintf('Cannot use %s method unless `wa72/htmlpagedom` is installed.', __METHOD__)
+            );
+        }
+    }
+
+    public function parse(string $html): string
```
Comment (mallardduck) [inline src/HtmlReplacer.php:25]: If we need to support full HTML docs and HTML fragments, then this method should:
1. Immediately determine if the input `$html` is a full DOM page, then
2. either use HtmlPage (used here) and work based on the Body, or
3. use the HtmlPageCrawler to parse the fragment.

Comment (Gummibeer) [reply src/HtmlReplacer.php:25]: I think that in PHP partial HTML is more common than a full document. Except you are implementing it as some kind of middleware to parse the whole HTML response.
But in general it should support both if possible.

Comment (mallardduck) [reply src/HtmlReplacer.php:25]: Addressed this by using the more general HTML parser, then adding a step where we check if the input HTML is a Page/Doc and selecting the `body` from that. As I was already replacing based on a HTML fragment (the body), supporting fragments as input was rather simple.

File: src/HtmlReplacer.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
+<?php
+
+namespace Astrotomic\Twemoji;
+
+use Astrotomic\Twemoji\Concerns\Configurable;
+use RuntimeException;
+use Wa72\HtmlPageDom\HtmlPageCrawler;
+
+/**
+ * @internal This class is marked as Internal as it is considered Experimental. Code subject to change until warning removed.
```
[the remark this answers is not in the document]
Comment (Gummibeer) [reply src/HtmlReplacer.php:10]: Yeah, `@internal` or `@deprecated` could be "aliases" for the experimental thing. Or even both. 😂

File: tests/Datasets/HtmlContent.php [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        <script type="text/javascript">
+            /* Ensure it won't touch emoji in script sections */
+            var badTime = "💩";
+        </script>
+    </body>
+</html>
+HTML,
+]);
+
+dataset('html-fragments', [
```
Comment (Gummibeer) [inline tests/Datasets/HtmlContent.php:114]: Could we add the most simple fragments here first? And add the more complex/combined ones to the end?

```html
<p>🚀</p>
```
```html
<img src="" alt="🎉"/>
```
```html
<a href="" title="🎈">Link ⛓️</a>
```

So some of these single-element examples.

I can also imagine that example which shouldn'T be replaced:

```html
<script>document.innerHTML = '🤷‍♂️';</script>
```

Comment (mallardduck) [reply tests/Datasets/HtmlContent.php]: Added those more simple fragments to the top and even moved the fragment dataset to be the "primary" source. 

This feedback clued me off to a flaw in my "early return" logic. So now I'm making sure I check for existence of "Text" nodes rather than children in general. Since your first example has no actual children but does have Text nodes I can work on that was being skipped with my original logic.

Only issue this highlighted is that currently the code does affect the script tag contents. I'll circle back later today/tomorrow to look into that aspect of things. Seems like I'll need to get creative about creating some sort of exclude list to skip Text nodes inside elements like `script`.

Comment (Gummibeer) [reply tests/Datasets/HtmlContent.php]: Same for `<style>` I think. But like said, this will be experimental - so It's also fine to add a `->wip()` `todo()` or whatever Pest test for such things. To highlight what has to be fixed/fulfilled to get it out of experimental mode.

How about passing Text only without any nodes?

```html
This is some fancy-💃 Markdown/WYSIWYG text with surrounding <p> tags disabled. 🎉
```

Or without surrounding elements but one somewhere in the middle?
```html
This is some fancy-💃 Markdown/WYSIWYG text with surrounding <code><p></code> tags disabled. 🎉
```

Comment (mallardduck) [reply tests/Datasets/HtmlContent.php]: So I haven't gotten back around to creating an exclude list concept, however I'm going to start taking that on next. Just wanted to report back on my findings related to passing those "plain text" examples. Unfortunately it seems that passing text not wrapped in a node is interpreted in a variety of ambiguous ways.

For instance, the first example technically should have those braces escaped as `&lt;` and `&gt;` - otherwise browsers (and PHP DOM) will automatically parse the P tag and add a close. So I've opted to test with the escaped codes and this line is interpreted as XML. So internally it's wrapped with a `_root` and as such the XPath to find the inner text nodes fails. Ultimately the result is the text gets returned untouched and Emoji's stay emojis. The second example is basically the same too - just without needing to escape the HTML tag since the `code` tag has a close.

Long story short, it got complicated really fast lol. So I think I need to take a step back and focus on the HTML parsing situation and get a handle on how various inputs are being parsed. The DOM library I picked for us uses Symfony/dom-crawler as it's core which in turn uses Mastermind\HTML5 library to compensate for PHP's core dom having some failings.

All that said, my next step will be to create my own test repo to assess the overall situation for PHP's HTML parsing abilities. Since I think we will want a more clear picture about what behaviour I'm seeing from each mechanic. For instance, how is PHP core `ext-dom` treating things, vs how `Mastermind\HTML5` parser, etc.

Comment (mallardduck) [reply tests/Datasets/HtmlContent.php]: More info here: https://github.com/mallardduck/php-html-parsing-report/actions/runs/3191495563/jobs/5207874729

https://github.com/mallardduck/php-html-parsing-report/tree/main/tests/__snapshots__

Comment (Gummibeer) [reply tests/Datasets/HtmlContent.php]: So in general I can say that the symfony Dom crawler is an outstanding package and one of if not the best solution for HTML in PHP.

Like said before: don't escalate this too far. With tests that are meant to fail we at least have a good list of things needed for a final release. And even if you fix some of those, take the chance for more PRs. 😉🎃

Regarding the plain text thing: how about a detector with some cases/match arms?
1. full HTML document
2. partial HTML but starting/ending with a tag or even surrounded by one
3. text with HTML sprinkles - solution could be to surround it with `<div>`, do the job and afterwards take `$div->innerHTML`

Again, these are just ideas - I'm also personally more happy to merge several smaller and focused PRs than one beast of a PR. As it's also easier to review these smaller ones.

Comment (mallardduck) [reply tests/Datasets/HtmlContent.php]: Thank you for that perspective! Always appreciate the sage advice and I think you nailed it. I'll start to refocus this PR to cover less of the edge cases and focus on the most common uses. Then I can adjust the tests a bit to give us a guiding star for future enhancements.

Comment (Gummibeer) [reply tests/Datasets/HtmlContent.php]: And one last thing: personally I would prefer datasets only for the same thing. Meaning really the same test case.

**it does not replace emojis in HTML attributes**
```html
<a title="🚀">hello</a>
<img alt="🖼️"/>
```

**it replaces emojis in text nodes**
```html
<p>🎉</p>
<div>💃</div>
```

So I'm more happy with a lot of copy'n'paste test cases with better/easier descriptions than one test case running through a massive dataset and no one really knows what's tested with that data-point and what is still missing or not covered. So every expectation to the lib should be one test case - and we can bullet proof that expectation by adding multiple variants of it. Like `<a>` and `` for the attributes. And later someone adds `<button>` to it.

Comment (mallardduck) [reply tests/Datasets/HtmlContent.php]: I've refactored the tests now and followed your advice of individual tests. Great idea as this allows me to more obviously target edge cases. So I've taken advantage of that by adding some explicit expectations to tests.

Comment (mallardduck) [reply tests/Datasets/HtmlContent.php]: Good news!



I've caught the root of the issue I've been chasing that I thought was something I did wrong. Specifically the way HTML attributes were being cast/converted to their HTML entities equivalent instead of staying unicode.

It turns out that this is caused by DOMDocument itself. I thought it was being caused by either: a) the xpath I found to use, or b) the HTML5 transforming library I opted to choose. In the end, reading the comments on PHP's docs for DOMDocument was my salvation! Specifically the [15 year old comment](https://www.php.net/manual/en/domdocument.loadhtml.php#78243) from `xuanbn` that said:

    I have discovered that, to help loadHTML() process utf file correctly, the meta tag should come first, before any utf string appear. For example, this HTML file. 

    <html>
        <head>
            <meta http-equiv="content-type" content="text/html; charset=utf-8">
            <title> Vietnamese - Tiếng Việt</title>
        </head>
    <body></body>
    </html>

With that new info I added an additional test that shows how the HTML being input determines how DOMDocument treats things up on calling `saveHtml()`. So I think this gets us to a lot better place where I've created a working solution, but we show that it's experimental and highlight the issue with a (skipped) failing test.

File: tests/Unit/HtmlReplacerPageTest.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    expect($results)->toContain('5🚀')->not()->toContain('&#');
+    assertMatchesTextSnapshot($results);
+})->skip('This will fail due to "incorrect" meta charset method, we need to consider how to address that.');
+
+it('will replace the Emoji on page, but not in head - NEW Charset Method', function () {
+    $pageHtml = <<<'HTML'
+<!DOCTYPE html>
+<html lang="en">
+    <head>
+    	<meta http-equiv="content-type" content="text/html; charset=utf-8">
```
Comment (mallardduck) [inline tests/Unit/HtmlReplacerPageTest.php:78]: This is the "correct" meta tag to set charset to UTF8.
This will allow `DOMDocument` to preserve the raw Emoji in the document.

File: tests/Unit/HtmlReplacerPageTest.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    expect($results)->toContain('5🚀')->not()->toContain('&#');
+    assertMatchesTextSnapshot($results);
+});
+
+it('will replace the Emoji on page, but not in head - OLD Charset Method', function () {
+    $pageHtml = <<<'HTML'
+<!DOCTYPE html>
+<html lang="en">
+    <head>
+        <meta charset="UTF-8">
```
Comment (mallardduck) [inline tests/Unit/HtmlReplacerPageTest.php:51]: This is the "incorrect" method to set charset to UTF8.
Unfortunately it will not be respected by `DOMDocument` and Emoji will become HTML entities.

File: tests/__snapshots__/HtmlReplacerPageTest__it_will_replace_the_Emoji_on_page,_but_not_in_head_-_NEW_Charset_Method__1.txt [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+        <meta http-equiv="X-UA-Compatible" content="ie=edge">
+        <title>HTML 5🚀 Boilerplate</title>
+        <link rel="stylesheet" href="style.css">
+    </head>
+    <body>
+        <h1>Do a quick kickflip! <img src="https://twemoji.maxcdn.com/v/latest/72x72/1f6f9.png" alt="🛹" width="72" height="72" loading="lazy" class="twemoji"></h1>
+        <p>This is HTML text that should be replaced, but the emoji in the head should not.</p>
+        <h2>Time for a CRAB RAVE!</h2>
+        <p><img src="https://twemoji.maxcdn.com/v/latest/72x72/1f980.png" alt="🦀" width="72" height="72" loading="lazy" class="twemoji"><img src="https://twemoji.maxcdn.com/v/latest/72x72/1f980.png" alt="🦀" width="72" height="72" loading="lazy" class="twemoji"><img src="https://twemoji.maxcdn.com/v/latest/72x72/1f980.png" alt="🦀" width="72" height="72" loading="lazy" class="twemoji"><img src="https://twemoji.maxcdn.com/v/latest/72x72/1f980.png" alt="🦀" width="72" height="72" loading="lazy" class="twemoji"><img src="https://twemoji.maxcdn.com/v/latest/72x72/1f980.png" alt="🦀" width="72" height="72" loading="lazy" class="twemoji"></p>
```
Comment (mallardduck) [inline tests/__snapshots__/HtmlReplacerPageTest__it_will_replace_the_Emoji_on_page,_but_not_in_head_-_NEW_Charset_Method__1.txt:13]: See here, how the `alt` HTML attribute is still a valid Emoji.

File: tests/__snapshots__/HtmlReplacerFragmentTest__it_can_handle_text_with_an_outer_P_tag__1.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+<p>This is some fancy-<img src="https://twemoji.maxcdn.com/v/latest/72x72/1f483.png" alt="&#128131;" width="72" height="72" loading="lazy" class="twemoji"> Markdown/WYSIWYG text with surrounding <p> tags enabled. <img src="https://twemoji.maxcdn.com/v/latest/72x72/1f389.png" alt="&#127881;" width="72" height="72" loading="lazy" class="twemoji"></p></p>
```
Comment (mallardduck) [inline tests/__snapshots__/HtmlReplacerFragmentTest__it_can_handle_text_with_an_outer_P_tag__1.txt:1]: However with this fragment over here it's obvious `DOMDocument` executes the normal behavior. The `alt` attribute is incorrectly encoded to html entity.

File: tests/Unit/HtmlReplacerPageTest.php (hunk 6 above)
Comment (mallardduck) [reply tests/Unit/HtmlReplacerPageTest.php:51]: I think this can be fixed before merge - we should just wrap all fragments in our own HTML page that uses the correct charset meta tag. This will ensure they are encoded correctly and then we can capture the fragment from our "template".

File: tests/Unit/HtmlReplacerPageTest.php (hunk 5 above)
Comment (Gummibeer) [reply tests/Unit/HtmlReplacerPageTest.php:78]: Will this `<meta charset="utf-8"/>` work as well? As that's what PHPstorm automatically creates and I think is the "modern" HTML5 way to do it. 🤔

**Comment by Gummibeer** [comment]
Just let me know when you want me to review and merge it^^

**Comment by mallardduck** [comment]
@Gummibeer - Hiya, Think that I've got things in a far more functional state now. The solution I've landed on seems to handle 80% of the cases via the main "happy path" - that is works perfect without a hitch. Then there's the ones with incorrect or missing meta `content-type` meta tags; these cases will work just fine, however will have the correct meta tag injected.
